### PR TITLE
Detect: noopt validation

### PR DESCRIPTION
**Pull request by giannitedesco**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: None.

Describe changes:

Rule parser does not currently validate that NOOPT options actually have options. This allows for a simple typo of ":" for ";" to lead to rule bugs. Two such bugs exist in latest ET-open ruleset. I added a patch to validate this.

While validating the change, discovered that detect-fast-pattern was marked NOOPT when it should be OPT_OPTIONAL.

#suricata-verify-pr:
#suricata-verify-repo:
#suricata-verify-branch:
#suricata-update-pr:
#suricata-update-repo:
#suricata-update-branch:
#libhtp-pr:
#libhtp-repo:
#libhtp-branch:

### CODE REVIEW

**Comment by victorjulien** [comment]
Hi @giannitedesco, I think this looks good, thanks. Did you sign the contrib agreement? I see no record of this on our side. Thanks!

**Comment by giannitedesco** [comment]
> Hi @giannitedesco, I think this looks good, thanks. Did you sign the contrib agreement? I see no record of this on our side. Thanks!

Woops, didn't see the form at the bottom. Done!

File: src/detect-parse.c [hunk 1]
```diff
@@ -708,6 +708,12 @@ static int SigParseOptions(DetectEngineCtx *de_ctx, Signature *s, char *optstr,
                     optname, optstr);
             goto error;
         }
+    } else if (st->flags & SIGMATCH_NOOPT) {
+        if (optvalue && strlen(optvalue)) {
+            SCLogError(SC_ERR_INVALID_SIGNATURE, "unexpected option to %s keyword: \'%s'", optname,
```
Comment (victorjulien) [inline src/detect-parse.c:713]: I don't think `'` needs to be escaped here?

**Comment by catenacyber** [comment]
Thanks got this. Could there be a unit test for this ? Which signature shows a different behavior ?

File: src/detect-parse.c (hunk 1 above)
Comment (giannitedesco) [reply src/detect-parse.c:713]: Yeah, weird, I dunno where that came from, will fix it

Comment (giannitedesco) [reply src/detect-parse.c]: Yeah, it came from copy and pasting the above check. Fixed it in both places.

**Comment by giannitedesco** [comment]
> Thanks got this. Could there be a unit test for this ? Which signature shows a different behavior ?

These 2 ET open rules:
```
alert http $HOME_NET any -> $EXTERNAL_NET any (msg:"ET MALWARE MSIL/Agent.TRM Data Exfil (sysinfo)"; flow:established,to_server; http.method; content:"POST"; http.start; content:"Cookie|3a 20|dkv="; fast_pattern; http.cookie; content:"dkv="; startswith; content:"|3b|YSC="; distance:32; within:5; pcre:"/^dkv=[a-f0-9]{32}\x3bYSC=\d+$/C"; http.header_names; content:!"Referer"; content:!"Content-Type"; http.request_body; content:"DQpIb3N0IE5hbWU6"; startswith: reference:md5,d2b81c4f5d075daa681f823cc9a5e4c0; reference:url,twitter.com/w3ndige/status/1247547923845578755; classtype:command-and-control; sid:2029855; rev:1; metadata:affected_product Windows_XP_Vista_7_8_10_Server_32_64_Bit, attack_target Client_Endpoint, created_at 2020_04_09, deployment Perimeter, deployment SSLDecrypt, former_category MALWARE, signature_severity Major, updated_at 2020_04_09;)
alert http $HOME_NET any -> $EXTERNAL_NET any (msg:"ET MALWARE IP Grabber CnC Activity"; flow:established,to_server; http.method; content:"POST"; http.uri; content:"/datarecord/"; endswith; http.request_body; content:"username="; startswith; content:"&content=IP%3a+"; distance:0; fast_pattern; content:"%0a"; endswith: reference:md5,635b08c141465abf86eaec88391b5ee6; classtype:command-and-control; sid:2030599; rev:2; metadata:affected_product Windows_XP_Vista_7_8_10_Server_32_64_Bit, attack_target Client_Endpoint, created_at 2020_07_27, deployment Perimeter, deployment SSLDecrypt, former_category MALWARE, performance_impact Low, signature_severity Major, updated_at 2020_07_27;)
```

Happy to add a test, but I might need a bit of guidance on how best to do that, the signature data structure looks quite complex.

**Comment by catenacyber** [comment]
You can add a test in file src/tests/detect-parse.c 

That means adding 
`    UtRegisterTest("DetectParseTest02", DetectParseTest02);` or find out a better name

Then define the function `DetectParseTest02 `

I guess it should be (with more comments) like
```
    DetectEngineCtx *de_ctx = DetectEngineCtxInit();
    FAIL_IF_NULL(de_ctx);

    Signature *sig = DetectEngineAppendSig(de_ctx,
            whatever signature");
    FAIL_IF_NOT_NULL(sig);

    DetectEngineCtxFree(de_ctx);
    PASS;
```

**Comment by giannitedesco** [comment]
>     Signature *sig = DetectEngineAppendSig(de_ctx,
>             whatever signature");
>     FAIL_IF_NOT_NULL(sig);

Ah, got it, just check that we barf, yeah, that makes sense. Can do.

**Comment by giannitedesco** [comment]
Regression test added. Merry xmas!

**Comment by jlucovsky** [comment]
Just a gentle reminder to create a new branch and PR instead of updating the original branch/pr directly. Some information is at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Github_work_flow

**Comment by victorjulien** [comment]
Merged in #5693, thanks!

Turns out there was an error in the ETpro set as well. Nice work :)

**Pull request closed without merging**